### DISCUSSION TITLE
Hover and go to definition not working inside a builtin library

### DISCUSSION THREAD

**Discussion by ydaveluy** [eclipse-langium/langium#1656] [General]
Hello,

By default, when following the [builtin library guide](https://langium.org/docs/recipes/builtin-library/), the hover and the go to definition is not working inside a builtin library.

In extension/main.ts, client options is defined (by langium CLI ?) as
```ts
// Options to control the language client
clientOptions: LanguageClientOptions = {
    documentSelector: [{ scheme: 'file', language: 'mydsl' }],
}
```

Defining the scheme to 'file' prevent hover and go to definition from working.
Removing the scheme option seems to solve this issue:
```ts
// Options to control the language client
clientOptions: LanguageClientOptions = {
    documentSelector: [{ language: 'mydsl' }],
}
```

1. Is it the proper way to fix this issue ?
2. What is the interest to define the scheme to 'file' by default in the langium CLI ?

**Comment by msujew** [comment]
Hey @ydaveluy,

I believe using only the `file` scheme is a recommendation of the VS Code docs. But I think we should actually change that. If you're interested, please contribute a PR :)

**Comment by ydaveluy** [comment]
It is also possible to register the language twice with 'file' and 'builtin' scheme like this:
```ts
// Options to control the language client
clientOptions: LanguageClientOptions = {
    documentSelector: [{ scheme: 'file', language: 'mydsl' }, { scheme: 'builtin', language: 'mydsl' }],
}
```

It is probably better if the scheme is recommended.

I can provide a PR to update https://github.com/eclipse-langium/langium-website/blob/4dfaa9b649a85a07ee845618f05e5b0c4dcfa9fe/hugo/content/docs/recipes/builtin-library.md

**Reply by msujew** [reply]
I wouldn't necessarily say so. `file` is used [here](https://code.visualstudio.com/api/language-extensions/language-server-extension-guide#explaining-the-language-client) in the vscode documentation and I've just adapted everything from there back in 2021. I personally believe most Langium projects are better served with specifiying the scheme as `'*'`. However, a PR to the website is nevertheless appreciated :)

**Reply by ydaveluy** [reply]
@msujew here is the link to the PR: https://github.com/eclipse-langium/langium-website/pull/252

Wouldn't it be wise to also modify the langium generator to handle all schemes by default? I can provide a PR for that as well.

**Reply by msujew** [reply]
Yes, a PR for the yeoman generator would be appreciated as well :)

**Comment by ydaveluy** [comment]
Hi @msujew,

I would like to warn you about a subtle bug I encountered when using
 ```ts
 documentSelector: [{ scheme: '*', language: 'mydsl' }],
```

I noticed that when using the diff utility from `Source Control` on mydsl file, the old revision is loaded inside LangiumDocments with a URI like `git:/path/to/file.mydsl`.
When a validator check for unique name, it creates lots of duplicates with the old revision. Even worse, sometime, some cross reference from other file are linked with the old revision instead of the local file. Closing the `Source Control` utility does not change anything. VSCode must be restarted to go back in a correct state.

For these reason I discourage to use '*' for the scheme and I encourage to explicitly define the "valid" schemes manually.

**Reply by msujew** [reply]
Oh, ok that's interesting. Good to know, thanks 👍 

It might make sense to revert the changes we did for the yeoman generator and adjust the documentation.

**Reply by ydaveluy** [reply]
Yes sure!

Here is the PR for the website : https://github.com/eclipse-langium/langium-website/pull/255
Do you need a PR for yeoman or can you revert the commit directly ?
